File: Python/ds/DiameterOfBinaryTree.py
'''
Theory-
The diameter of a tree can be defined as the
maximum distance between two leaf nodes.
Here, the distance is measured in terms of the
total number of nodes present along the path
of the two leaf nodes, including both the leaves.
'''


import queue
from sys import stdin


class Node:
    def __init__(self, data):
        self.data = data
        self.left = None
        self.right = None


def levelOrderInput():
    '''
    Summary Line:
    This function helps us to take level order
    input and build the tree.

    Args:
    No arguments

    Returns:
    Root- root of the tree formed using level
    order input
    '''
    tlist = list(map(int, stdin.readline().strip().split(" ")))
    start = 0
    data = tlist[start]
    start += 1
    if data != -1:
        root = Node(data)
    else:
        return None
    q = queue.Queue()
    q.put(root)
    while not q.empty():
        current = q.get()
        data = tlist[start]
        start += 1
        if data != -1:
            leftChild = Node(data)
            current.left = leftChild
            q.put(leftChild)
        data = tlist[start]
        start += 1
        if data != -1:
            rightChild = Node(data)
            current.right = rightChild
            q.put(rightChild)
    return root


def Diameter(root):
    '''
    Summary Line:
    This function helps us to calculate the diameter
    of our tree. Diameter is the number of nodes
    between the two farthest nodes in our tree.

    Args:
    root- root of our inputted tree

    Returns-
    h- height of the tree
    d- diameter of the tree

    Intuition-
    Height is calculated because the diameter
    is the sum of max left and max right heights.
    '''
    if root is None:
        return 0, 0
    lh, ld = Diameter(root.left)
    rh, rd = Diameter(root.right)
    h = 1 + max(lh, rh)
    d = lh + rh
    b = max(d, max(ld, rd))
    return h, b

if __name__ == "__main__":
    print("Enter level order input for the tree:")
    root = levelOrderInput()
    h, d = Diameter(root)
    print(d + 1)  # added 1 for the root node


'''

Input format-
The first and the only line of input will
contain the node data, all separated by a single space.
Since -1 is used as an indication whether the
left or right node data exist for root,
it will not be a part of the node data.

Sample Input-
Enter level order input for the tree:
1 2 3 4 5 6 7 -1 -1 -1 -1 -1 -1 -1 -1

Illustration for the tree formed using the input-
     1
   /   \
  2     3
 / \   / \
4   5 6   7

Sample Output-
5

Time complexity- O(N^2)

'''

File: Java/ds/Diameter_of_Binary_tree.java
//  The function is expected to return the number of edges between two nodes which are farthest from each other in terms of edges


import java.io.*;
import java.util.*;

public class Main {
  public static class Node {          // Node Class
    int data;
    Node left;
    Node right;

    Node(int data, Node left, Node right) {
      this.data = data;
      this.left = left;
      this.right = right;
    }
  }

  public static class Pair {
    Node node;
    int state;

    Pair(Node node, int state) {
      this.node = node;
      this.state = state;
    }
  }

  public static Node construct(Integer[] arr) {       // Function for Conteruction of Binary Tree
    Node root = new Node(arr[0], null, null);
    Pair rtp = new Pair(root, 1);

    Stack<Pair> st = new Stack<>();
    st.push(rtp);

    int idx = 0;
    while (st.size() > 0) {
      Pair top = st.peek();
      if (top.state == 1) {
        idx++;
        if (arr[idx] != null) {
          top.node.left = new Node(arr[idx], null, null);
          Pair lp = new Pair(top.node.left, 1);
          st.push(lp);
        } else {
          top.node.left = null;
        }

        top.state++;
      } else if (top.state == 2) {
        idx++;
        if (arr[idx] != null) {
          top.node.right = new Node(arr[idx], null, null);
          Pair rp = new Pair(top.node.right, 1);
          st.push(rp);
        } else {
          top.node.right = null;
        }

        top.state++;
      } else {
        st.pop();
      }
    }

    return root;
  }

  public static void display(Node node) {
    if (node == null) {
      return;
    }

    String str = "";
    str += node.left == null ? "." : node.left.data + "";
    str += " <- " + node.data + " -> ";
    str += node.right == null ? "." : node.right.data + "";
    System.out.println(str);

    display(node.left);
    display(node.right);
  }

  
  public static class DiaSolver   // Class to help us return the height and diameter of tree to ease Calculation
  {
      int h;
      int dia;
      DiaSolver(int h,int dia)
      {
          this.h=h;
          this.dia=dia;
      }
      
  }

  public static DiaSolver diameter1(Node node) {                  //   Function to calculate Diameter
    if(node==null)
    {
        return new DiaSolver(-1,0);  // if node is null then its height is -1 and diameter is 0
    }
    
    DiaSolver l = diameter1(node.left);   // fuction calling left part
    
    DiaSolver r = diameter1(node.right);  // function calling right part
    
    int d = l.h + r.h + 2;                    // the current diameter is left height + right height + 2
    
    int cdia = Math.max(d,Math.max(l.dia,r.dia)); // calculating current maximum diameter
    
    int ch=Math.max(l.h,r.h)+1;                   // calculating height
    
    return new DiaSolver(ch,cdia);               // return ans in form of DiaSolver 
  }

  public static void main(String[] args) throws Exception {
    BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    int n = Integer.parseInt(br.readLine());
    Integer[] arr = new Integer[n];
    String[] values = br.readLine().split(" ");
    for (int i = 0; i < n; i++) {
      if (values[i].equals("n") == false) {
        arr[i] = Integer.parseInt(values[i]);
      } else {
        arr[i] = null;
      }
    }

    Node root = construct(arr);

    int diameter = 0;
    DiaSolver  res = diameter1(root); // function calling
    System.out.println(res.dia);     // prining maximum diameter 
  }

}

/*
Sample Input/Output:
Input: 19
       50 25 12 n n 37 30 n n n 75 62 n 70 n n 87 n n
Output: 6
Time Complexity: O(n) 
*/

/*
Sample Input/Output:
Input: 11
       1 2 4 n n 5 n n 3 n n
Output: 3
Time Complexity: O(n) 
*/
